In [6]:
import os 
import sys
import pandas as pd
import pymongo
import configparser as cfg

# Load csv file into database
### Change the file name as appropriate

In [71]:
csv_file = "//NAS-Miftek/Public/Data Captures/M2DemoPlatform/valery_files/Recording1/20241220 DGMc 3 7 10 brt yellow recording run 3 laser 5k events.csv"
channel_info = "C:/Users/MichaelKolb/Github/granite/granite01/data/config/channelConfig_M2_UsingABCDMb.ini"

# Check if the files exist if not exit
if not os.path.exists(csv_file):
    print("Error:  File (csv) does not exist")
    exit(1)

if not os.path.exists(channel_info):
    print("Error:  File (config) does not exist")
    exit(1)

cursor = cfg.ConfigParser(strict=False)
cursor.read(channel_info)
channel_config = cursor.items('adc_channels')
print(channel_config)
print(type(channel_config))
channel_names = [i[1] for i in channel_config]
print(channel_names)

[('daq0.0', '360nm'), ('daq0.1', '382nm'), ('daq0.2', '404nm'), ('daq0.3', '426nm'), ('daq0.4', '448nm'), ('daq0.5', '470nm'), ('daq0.6', 'NC'), ('daq0.7', '491nm'), ('daq0.8', '513nm'), ('daq0.9', '535nm'), ('daq0.10', '557nm'), ('daq0.11', '579nm'), ('daq1.0', '590nm'), ('daq1.1', '568nm'), ('daq1.2', '546nm'), ('daq1.3', '524nm'), ('daq1.4', '502nm'), ('daq1.5', '480nm'), ('daq1.6', 'NC'), ('daq1.7', '459nm'), ('daq1.8', '437nm'), ('daq1.9', '415nm'), ('daq1.10', '393nm'), ('daq1.11', '371nm'), ('daq2.0', 'NC'), ('daq2.1', '601nm'), ('daq2.2', '623nm'), ('daq2.3', '645nm'), ('daq2.4', '667nm'), ('daq2.5', '689nm'), ('daq2.6', 'NC'), ('daq2.7', '710nm'), ('daq2.8', '732nm'), ('daq2.9', '754nm'), ('daq2.10', '776nm'), ('daq2.11', '798nm'), ('daq3.0', 'NC'), ('daq3.1', '809nm'), ('daq3.2', '787nm'), ('daq3.3', '765nm'), ('daq3.4', '743nm'), ('daq3.5', '721nm'), ('daq3.6', 'NC'), ('daq3.7', '700nm'), ('daq3.8', '678nm'), ('daq3.9', '656nm'), ('daq3.10', '634nm'), ('daq3.11', '612nm'), (

In [72]:
software_names = ['laser1_adc_peak', 'laser1_adc_area', 'laser1_adc_width', 'laser1_photon_total',
                  'laser2_adc_peak', 'laser2_adc_area', 'laser2_adc_width', 'laser2_photon_total',
                  'laser3_adc_peak', 'laser3_adc_area', 'laser3_adc_width', 'laser3_photon_total',
                  'laser4_adc_peak', 'laser4_adc_area', 'laser4_adc_width', 'laser4_photon_total',
                  'laser5_adc_peak', 'laser5_adc_area', 'laser5_adc_width', 'laser5_photon_total',
                  'laser6_adc_peak', 'laser6_adc_area', 'laser6_adc_width', 'laser6_photon_total',
                  'laser7_adc_peak', 'laser7_adc_area', 'laser7_adc_width', 'laser7_photon_total',
                  'time']

hardware_names = ['laser1_abs_area_hw', 'laser1_pos_peak_hw', 'laser1_pos_width_hw', 'laser1_neg_peak_hw', 'laser1_neg_width_hw',
                  'laser2_abs_area_hw', 'laser2_pos_peak_hw', 'laser2_pos_width_hw', 'laser2_neg_peak_hw', 'laser2_neg_width_hw',
                  'laser3_abs_area_hw', 'laser3_pos_peak_hw', 'laser3_pos_width_hw', 'laser3_neg_peak_hw', 'laser3_neg_width_hw',
                  'laser4_abs_area_hw', 'laser4_pos_peak_hw', 'laser4_pos_width_hw', 'laser4_neg_peak_hw', 'laser4_neg_width_hw',
                  'laser5_abs_area_hw', 'laser5_pos_peak_hw', 'laser5_pos_width_hw', 'laser5_neg_peak_hw', 'laser5_neg_width_hw',
                  'laser6_abs_area_hw', 'laser6_pos_peak_hw', 'laser6_pos_width_hw', 'laser6_neg_peak_hw', 'laser6_neg_width_hw',
                  'laser7_abs_area_hw', 'laser7_pos_peak_hw', 'laser7_pos_width_hw', 'laser7_neg_peak_hw', 'laser7_neg_width_hw',
                  'time_hw']


collections_software = {}
collections_hardware = {}
db_client = None
db_software = None
db_hardware = None

In [73]:
# def setup_database():
# """
# @brief Function that is used to setup the database

# @details This method will setup the database connection and the collections that are used 
#         in the database. This method will also clear the database of any data that is
#         currently in the database.
        
# @return void
# """
db_client = pymongo.MongoClient("mongodb://localhost:27017/")
db_software = db_client["software"]
db_hardware = db_client["hardware"]

names = db_software.list_collection_names()
collections_software = {}
for name in names:
    db_software.drop_collection(name)
    
try:
    print("Software db: ", db_software)
    for name in software_names:
        collections_software[f'{name}'] = db_software.create_collection(name)
except pymongo.errors.CollectionInvalid:
    print(f"\nException caught in database setup: {e}: {type(e)}. \nSHUTTING DOWN!\n")
    sys.exit()
print("Collections software: ",collections_software)
    
names = db_hardware.list_collection_names()
collections_hardware = {}
for name in names:
    db_hardware.drop_collection(name)
try:
    for name in hardware_names:
        collections_hardware[f'{name}'] = db_hardware.create_collection(name)
except Exception as e:
    print(f"\nException caught in database setup: {e}: {type(e)}. \nSHUTTING DOWN!\n")
    sys.exit()

Software db:  Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'software')
Collections software:  {'laser1_adc_peak': Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'software'), 'laser1_adc_peak'), 'laser1_adc_area': Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'software'), 'laser1_adc_area'), 'laser1_adc_width': Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'software'), 'laser1_adc_width'), 'laser1_photon_total': Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'software'), 'laser1_photon_total'), 'laser2_adc_peak': Collection(Database(MongoClient(host=['localhost:27017'], document_class=dict, tz_aware=False, connect=True), 'software'), 'laser2_adc_peak'), 'laser2_adc_area': Collection(D

In [ ]:
# def load_database(filename):
# """
# @brief Function that is used to load the database from a file

# @details This method will load the database from a file. The file that the database is 
#         loaded from is the filename that is passed to the method. This method will also
#         check to see if the collections that are in the file are in the database. If they 
#         are not in the database, the method will not load the database.
        
# @param filename: The filename that the database is loaded from

# @return void
# """
df = pd.read_csv(f"{csv_file}", low_memory=False)
df = df.drop(df.__len__()-1)

last_line = ''
with open(f"{csv_file}", "r") as file:
    for line in file:
        pass
    last_line = line
    file.close()
    
last_line = last_line.split(" ")
total = int(last_line[0])
load_names = last_line[1:]

load_names2 = []
total2 = None
for i in range(1,len(last_line),1):
    if last_line[i].isdigit():
        load_names2 = last_line[i:]
        break

df2 = None
if load_names2:
    load_names = last_line[1:i]
    total2 = int(load_names2[0])
    load_names2 = load_names2[1:]
    split = total * len(load_names) + len(load_names)
    df2 = df.iloc[split+1:]
    df = df.iloc[:split-1]

# ? Maybe combine the database loads
for i in range(len(load_names)):
    if load_names[i] not in list(collections_software.keys()):
        print("File to load has collections that do not exist in this file, unable to load: ", load_names[i], " not in ", list(collections_software.keys()))
        sys.exit() 
    if (load_names[i] != 'time') and ('laser' not in load_names[i]):
        load_names[i] = 'laser1_' + load_names[i]

    if i == 0:
        continue
    df = df.drop(total*i + (i-1))

for i in range(len(df.columns)):
    if i == 0:
        continue
    elif i == 1:
        df = df.astype({df.columns[i]: int})
    else:
        df = df.astype({df.columns[i]: float})

#NOTE:  This insert into the database walks through the different collections and inserts 
#       the data. The dataframes that correspond to the collections are created by slicing 
#       and saving as tempdf, then all NaNs are dropped from the collection and made into a 
#       dictionary. This dictionary is then inserted into the database.
for i in range(len(load_names)):
    tempdf = df.iloc[(total*i):((total*(i+1))-1)]
    temp = tempdf.dropna(axis=1).to_dict('records')
    collections_software[load_names[i]].insert_many(temp)

if total2:
    names_exist = []
    for i in range(len(load_names2)):
        if load_names2[i] in list(collections_hardware.keys()):
            #NOTE:  Instead of panicking, just skip the collection and move on. If there are 
            #       no collections to load, then return
            names_exist.append(i)
            continue

        if not names_exist:
            print("File to load has collections (hardware) that do not exist, unable to load")
            sys.exit()
        
        if i == 0:
            continue
        df2 = df2.drop(split + total2*i + (i-1))

    for i in range(len(df2.columns)):
        if i == 0:
            continue
        elif i == 1:
            df2 = df2.astype({df.columns[i]: int})
        else:
            df2 = df2.astype({df.columns[i]: float})

    for i in names_exist:
        tempdf = df2.iloc[(total2*i):((total2*(i+1))-1)]
        temp = tempdf.dropna(axis=1).to_dict('records')
        collections_hardware[load_names2[i]].insert_many(temp)

print("Database Loaded")

['laser1_adc_width', 'laser1_adc_area', 'laser2_photon_total', 'laser3_adc_peak', 'laser2_adc_width', 'laser3_photon_total', 'laser2_adc_peak', 'laser2_adc_area', 'laser3_adc_width', 'time', 'laser1_adc_peak', 'laser1_photon_total', 'laser3_adc_area']
Database Loaded


In [121]:
# def close_database():
# """
# @brief Function that is used to close the database connection

# @details This method will close the database connection and the database worker. This
#         method will also drop all of the collections that are in the database. This will
#         clear the database of any data that is currently in the database.
    
# @return void
# """
try:
    names = db_software.list_collection_names()
    for name in names:
        db_software.drop_collection(name)
        
    names = db_hardware.list_collection_names()
    for name in names:
        db_hardware.drop_collection(name)
        
    db_client.close()
    db_client = None
    db_software = None
    db_hardware = None
except Exception as e:
    print(f"Error closing database connection: {e}")

# Information Manipulation

These following items will take all the information in the database and align it in the following format

## Laser (#)
### - Channel name (Wavelength nm) - W / Channel name (Wavelength nm) - H / Channel name (Wavelength nm) - A / Channel name (Wavelength nm) - C
#### - Event (Start - Finish)

This will happen for all number of lasers and then can be saved

In [75]:
def sort_laser_params(params):
    def sort_key(item):
        # Split by underscore and extract number and parameter
        laser_num = int(item.split('_')[0].replace('laser', ''))
        param = item.split('_')[1]
        return (laser_num, param)
    
    return sorted(params, key=sort_key)

soft_names = []

names = db_software.list_collection_names()
i = 0
for name in names:
    if db_software.get_collection(name).count_documents({}) == 0:
        continue

    soft_names.append(name)

if 'time' in soft_names:
    soft_names.remove('time')

# Sort and print the result
soft_names = sort_laser_params(soft_names)
print(soft_names)

['laser1_adc_width', 'laser1_adc_area', 'laser1_adc_peak', 'laser1_photon_total', 'laser2_adc_width', 'laser2_adc_peak', 'laser2_adc_area', 'laser2_photon_total', 'laser3_adc_width', 'laser3_adc_area', 'laser3_adc_peak', 'laser3_photon_total']


In [86]:
items = {}
for idx in range(soft_names.__len__()):
    cursor = db_software.get_collection(soft_names[idx]).find({}).sort('event', pymongo.ASCENDING)
    items[soft_names[idx]] = pd.DataFrame(list(cursor))
    items[soft_names[idx]].drop(columns=['_id', 'event'], inplace=True)

In [88]:
lst = list(items.keys())
print(lst[0])
items[lst[0]].head(5)

laser1_adc_width


ch 0   ch 1   ch 2   ch 3   ch 4   ch 5   ch 6   ch 7   ch 8   ch 9  ...  \
0  394.0  900.0  893.0  448.0  407.0  379.0  112.0  264.0  334.0  377.0  ...   
1  443.0  898.0  830.0  524.0  408.0  425.0  216.0  258.0  412.0  439.0  ...   
2  412.0  884.0  845.0  471.0  408.0  433.0  150.0  277.0  397.0  388.0  ...   
3  441.0  891.0  858.0  534.0  465.0  379.0   54.0  278.0  384.0  410.0  ...   
4  400.0  898.0  852.0  572.0  465.0  417.0  147.0  291.0  404.0  392.0  ...   

   ch 50  ch 51  ch 52  ch 53  ch 54  ch 55  ch 56  ch 57  ch 58  ch 59  
0  813.0  433.0    0.0    0.0  573.0    0.0  460.0  900.0    4.0    0.0  
1  815.0  459.0    0.0    0.0  660.0    0.0  400.0  900.0   51.0    0.0  
2  797.0  450.0    0.0    0.0  538.0    0.0  454.0  900.0   16.0    0.0  
3  891.0  844.0    0.0    0.0  832.0    0.0  110.0  900.0  575.0    0.0  
4  896.0  796.0    0.0    0.0  852.0    0.0  206.0  900.0  596.0    0.0  

[5 rows x 60 columns]

In [94]:
lst = list(items.keys())
print(len(items[lst[0]].columns))
print(len(channel_names))
print(len([ch for ch in channel_names]))

60
60
60


In [96]:
for key in items.keys():
    if 'area' in key:
        sigType = 'A'
    elif 'peak' in key:
        sigType = 'H'
    elif 'width' in key:
        sigType = 'W'
    elif 'photon' in key:
        sigType = 'P'
        items[key][[f'ch {i}' for i in range(48, 60)]] = 0
    items[key].columns = [f'{ch}-{sigType}' for ch in channel_names]

In [114]:
numLasers = 0
lasers = []
for key in items.keys():
    if key.split('_')[0] not in lasers:
        numLasers += 1
        lasers.append(key.split('_')[0])

print(numLasers)

lasers = []
for i in range(numLasers):
    keys = [key for key in items.keys() if f'laser{i+1}' in key]
    print(keys)
    lasers.append(
        pd.concat([items[key] for key in keys], axis=1)
    )

3
['laser1_adc_width', 'laser1_adc_area', 'laser1_adc_peak', 'laser1_photon_total']
['laser2_adc_width', 'laser2_adc_peak', 'laser2_adc_area', 'laser2_photon_total']
['laser3_adc_width', 'laser3_adc_area', 'laser3_adc_peak', 'laser3_photon_total']


In [115]:
lasers[0].head(5)

360nm-W  382nm-W  404nm-W  426nm-W  448nm-W  470nm-W   NC-W  491nm-W  \
0    394.0    900.0    893.0    448.0    407.0    379.0  112.0    264.0   
1    443.0    898.0    830.0    524.0    408.0    425.0  216.0    258.0   
2    412.0    884.0    845.0    471.0    408.0    433.0  150.0    277.0   
3    441.0    891.0    858.0    534.0    465.0    379.0   54.0    278.0   
4    400.0    898.0    852.0    572.0    465.0    417.0  147.0    291.0   

   513nm-W  535nm-W  ...  QPD BL-P  QPD BR-P  NC-P  NC-P  QPD Sum-P  QPD V-P  \
0    334.0    377.0  ...         0         0     0     0          0        0   
1    412.0    439.0  ...         0         0     0     0          0        0   
2    397.0    388.0  ...         0         0     0     0          0        0   
3    384.0    410.0  ...         0         0     0     0          0        0   
4    404.0    392.0  ...         0         0     0     0          0        0   

   QPD H-P  QPD S Baseline-P  QPD S Baselined-P  NC-P  
0        0                 0                  0     0  
1        0                 0                  0     0  
2        0                 0                  0     0  
3        0                 0                  0     0  
4        0                 0                  0     0  

[5 rows x 240 columns]

In [116]:
organized_cols = []
for ch in channel_names:
    for sign in ['A', 'H', 'W', 'P']:
        organized_cols.append(f'{ch}-{sign}')
print(organized_cols)

['360nm-A', '360nm-H', '360nm-W', '360nm-P', '382nm-A', '382nm-H', '382nm-W', '382nm-P', '404nm-A', '404nm-H', '404nm-W', '404nm-P', '426nm-A', '426nm-H', '426nm-W', '426nm-P', '448nm-A', '448nm-H', '448nm-W', '448nm-P', '470nm-A', '470nm-H', '470nm-W', '470nm-P', 'NC-A', 'NC-H', 'NC-W', 'NC-P', '491nm-A', '491nm-H', '491nm-W', '491nm-P', '513nm-A', '513nm-H', '513nm-W', '513nm-P', '535nm-A', '535nm-H', '535nm-W', '535nm-P', '557nm-A', '557nm-H', '557nm-W', '557nm-P', '579nm-A', '579nm-H', '579nm-W', '579nm-P', '590nm-A', '590nm-H', '590nm-W', '590nm-P', '568nm-A', '568nm-H', '568nm-W', '568nm-P', '546nm-A', '546nm-H', '546nm-W', '546nm-P', '524nm-A', '524nm-H', '524nm-W', '524nm-P', '502nm-A', '502nm-H', '502nm-W', '502nm-P', '480nm-A', '480nm-H', '480nm-W', '480nm-P', 'NC-A', 'NC-H', 'NC-W', 'NC-P', '459nm-A', '459nm-H', '459nm-W', '459nm-P', '437nm-A', '437nm-H', '437nm-W', '437nm-P', '415nm-A', '415nm-H', '415nm-W', '415nm-P', '393nm-A', '393nm-H', '393nm-W', '393nm-P', '371nm-A', 

In [118]:
for i in range(numLasers):
    lasers[i] = lasers[i][organized_cols]

In [119]:
lasers[0].head(5)

360nm-A  360nm-H  360nm-W  360nm-P     382nm-A  382nm-H  382nm-W  382nm-P  \
0  10253.0    237.0    394.0     23.0   5909052.0  13290.0    900.0   4929.0   
1  11357.0    332.0    443.0     20.0   7387968.0  16687.0    898.0   4643.0   
2  10241.0    257.0    412.0     19.0   8180508.0  19753.0    884.0   4805.0   
3  18953.0    402.0    441.0     62.0  16418637.0  30985.0    891.0   5413.0   
4  14245.0    276.0    400.0     47.0  15265750.0  31016.0    898.0   5082.0   

     404nm-A  404nm-H  ...  NC-W  NC-P  NC-P  NC-P    NC-P  NC-P  NC-P  NC-P  \
0   835824.0   3302.0  ...   0.0   0.0   0.0   0.0  5141.0   0.0   0.0     0   
1   846152.0   2837.0  ...   0.0   0.0   0.0   0.0  4648.0   0.0   0.0     0   
2   977108.0   2772.0  ...   0.0   0.0   0.0   0.0  5010.0   0.0   0.0     0   
3  2321086.0   5940.0  ...   0.0   0.0   0.0   0.0  5559.0   0.0   0.0     0   
4  2286748.0   5457.0  ...   0.0   0.0   0.0   0.0  5423.0   0.0   0.0     0   

   NC-P  NC-P  
0     0     0  
1     0     0  
2     0     0  
3     0     0  
4     0     0  

[5 rows x 528 columns]

In [120]:
for i in range(numLasers):
    lasers[i].to_csv(f'laser{i+1}.csv', index=False)

In [ ]:
# hard_names = []
# HARDWARE DB will work on this later          
# names = db_hardware.list_collection_names()
# for name in names:
#     if db_hardware.get_collection(name).count_documents({}) == 0:
#         continue
#     cursor = db_hardware.get_collection(name).find({}).sort('event', pymongo.ASCENDING)

#     df = pd.DataFrame(list(cursor))
    
#     hard_names.append(name)

Output of the below statement
Laser0  Value0 Laser1  Value1 Laser2  Value2
0     L1     120     L2     220     L3     320
1     L1     150     L2     250     L3     350
2     L1     180     L2     280     L3     380